File: 백준/Gold/14438. 수열과 쿼리 17/수열과 쿼리 17.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	N = Integer.parseInt(br.readLine());
    	arr = new int[N+1];
    	st = new StringTokenizer(br.readLine());
    	for(int i = 1; i<=N; i++) {
    		arr[i] = Integer.parseInt(st.nextToken());
    	}
    	
    	SegTree seg = new SegTree(N);
    	
    	M = Integer.parseInt(br.readLine());
    	StringBuilder sb = new StringBuilder();
    	for(int i = 0; i < M ;i++) {
    		st = new StringTokenizer(br.readLine());
    		int cmd = Integer.parseInt(st.nextToken());
    		
    		if(cmd == 1) {
    			int idx = Integer.parseInt(st.nextToken());
    			int value = Integer.parseInt(st.nextToken());
    			seg.update(1, 1, N, value, idx, idx);
    		}else if(cmd == 2) {
    			int qLeft = Integer.parseInt(st.nextToken());
    			int qRight = Integer.parseInt(st.nextToken());
    			sb.append(seg.query(1, 1, N,qLeft,qRight)).append('\n');
    		}
    	}
    	System.out.println(sb);
    }
    static int N,M,arr[];
    static int INF = Integer.MAX_VALUE / 2;
    static StringTokenizer st;
    static class SegTree{
    	int tree[], lazy[];
    	boolean isLazy[];
    	
    	public SegTree(int n) {
    		int h = (int)Math.ceil(Math.log(n) / Math.log(2));
    		int size = 1<<(h+1);
    		tree = new int[size];
    		lazy = new int[size];
    		isLazy = new boolean[size];
    		
    		this.build(1, 1, n);
    	}
    	private void build(int cur,int left, int right) {
    		if(left == right) {
    			tree[cur] = arr[left];
    			return;
    		}
    		int mid = (left + right) / 2;
    		this.build(cur * 2, left, mid);
    		this.build(cur * 2 + 1, mid + 1, right);
    		
    		tree[cur] = Math.min(tree[cur*2], tree[cur*2+1]);
    	}
    	
    	public void update(int cur, int left, int right, int value, int qLeft, int qRight) {
    		propagate(cur, left, right);
    		if(right < qLeft || qRight < left)
    			return;
    		
    		if(qLeft <= left && right <= qRight) {
    			isLazy[cur] = true;
    			lazy[cur] = value;
    			propagate(cur, left, right);
    			return;
    		}
    		
    		int mid = (left + right ) / 2;
    		update(cur * 2, left, mid, value, qLeft, qRight);
    		update(cur * 2 + 1, mid + 1, right ,value, qLeft, qRight);
    		
    		tree[cur] = Math.min(tree[cur * 2], tree[cur * 2 + 1]);
    		
    	}
    	
    	private void propagate(int cur, int left, int right) {
    		if (!isLazy[cur])
    			return;
    		
    		if(left != right) {
    			lazy[cur * 2] = lazy[cur];
    			lazy[cur * 2 + 1] = lazy[cur];
    			isLazy[cur * 2] = true;
    			isLazy[cur * 2 + 1] = true;
    		}
    		
    		tree[cur] = lazy[cur] * (right - left + 1);
    		isLazy[cur] = false;
    		lazy[cur] = INF;
    	}
    	
    	public int query(int cur, int left, int right, int qLeft, int qRight) {
       		propagate(cur, left, right);
    		if(right < qLeft || qRight < left)
    			return INF;
       		
       		if(qLeft <= left && right <= qRight)
       			return tree[cur];
       		
       		int mid = (left + right ) / 2;
       		int leftValue = query(cur * 2, left, mid ,qLeft, qRight);
       		int rightValue = query(cur * 2 + 1, mid + 1, right ,qLeft, qRight);
       		
       		return Math.min(leftValue, rightValue);
    	}
    }
}

File: 백준/Gold/14438. 수열과 쿼리 17/수열과 쿼리 17.py
import sys

def init(node, start, end):
    """세그먼트 트리 초기화"""
    if start == end:
        tree[node] = arr[start]
        return tree[node]

    mid = (start + end) // 2
    tree[node] = min(init(node * 2, start, mid), init(node * 2 + 1, mid + 1, end))
    return tree[node]


def query(node, start, end, left, right):
    """최솟값 쿼리"""
    if left > end or right < start:
        return float('inf')

    if left <= start and end <= right:
        return tree[node]

    mid = (start + end) // 2
    return min(query(node * 2, start, mid, left, right),
               query(node * 2 + 1, mid + 1, end, left, right))


def update(node, start, end, index, diff):
    """값 업데이트"""
    if index < start or index > end:
        return

    if start == end:  # 리프 노드
        tree[node] = diff # diff가 아닌 업데이트된 값 자체를 넣어야함
        return

    mid = (start + end) // 2
    update(node * 2, start, mid, index, diff)
    update(node * 2 + 1, mid + 1, end, index, diff)
    tree[node] = min(tree[node * 2], tree[node * 2 + 1]) # 부모 노드 업데이트


if __name__ == "__main__":
    input = sys.stdin.readline
    N = int(input())
    arr = list(map(int, input().split()))

    # 트리 크기 계산 (4*N으로 해도 충분)
    h = 1
    while (1 << h) < N:
        h += 1
    tree_size = 1 << (h + 1)
    tree = [0] * tree_size

    init(1, 0, N - 1)  # 세그먼트 트리 초기화

    M = int(input())
    for _ in range(M):
        a, b, c = map(int, input().split())
        if a == 1:
            # b번째 수를 c로 바꾼다
            # arr[b - 1] = c  # 배열 업데이트는 필요 없음 (세그먼트 트리에서만)
            update(1, 0, N - 1, b - 1, c)
        else:
            # b번째 수부터 c번째 수까지의 최솟값 출력
            print(query(1, 0, N - 1, b - 1, c - 1))